### PR TITLE
[Qol] Driver toolchain validate args

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
This is a clean up of TODO from #26136 

I'm not really sure if `validateSearchPathArgs` is darwin specific. 
But the other ones not moved, I think are for all platforms.

cc @jrose-apple @brentdax

### CODE REVIEW

File: lib/Driver/DarwinToolChains.cpp [hunk 1]
```diff
@@ -588,8 +588,66 @@ bool toolchains::Darwin::shouldStoreInvocationInDebugInfo() const {
...
+
+void
+toolchains::Darwin::validateDeploymentTarget(DiagnosticEngine &diags,
+                                             const llvm::opt::ArgList &args) const {
+  const Arg *A = args.getLastArg(options::OPT_target);
+  if (!A)
+    return;
+  
+  // Check minimum supported OS versions.
+  llvm::Triple triple(llvm::Triple::normalize(A->getValue()));
```
Comment (jrose-apple) [inline lib/Driver/DarwinToolChains.cpp]: Now that this is in the toolchain you can actually do better: `getTriple()` gets the already-computed triple.

File: lib/Driver/DarwinToolChains.cpp [hunk 2]
```diff
@@ -588,8 +588,66 @@ bool toolchains::Darwin::shouldStoreInvocationInDebugInfo() const {
...
+    if (triple.isOSVersionLT(2, 0)) {
+      diags.diagnose(SourceLoc(), diag::error_os_minimum_deployment,
+                     "watchOS 2.0");
+      return;
+    }
+  }
+}
+
+void
+toolchains::Darwin::validateBridgingHeaderArgs(DiagnosticEngine &diags,
```
Comment (jrose-apple) [inline lib/Driver/DarwinToolChains.cpp]: We don't have a good story for these flags on non-Apple platforms, but if I'm remembering correctly they're currently accepted and therefore these conditions should probably still be enforced everywhere.

**Review by jrose-apple — COMMENTED** [review]
Thanks for finishing this up!

File: lib/Driver/ToolChains.h [hunk 3]
```diff
@@ -40,6 +40,12 @@ class LLVM_LIBRARY_VISIBILITY Darwin : public ToolChain {
 
   bool shouldStoreInvocationInDebugInfo() const override;
 
+private:
```
Comment (brentdax) [inline lib/Driver/ToolChains.h]: If these are never called outside of DarwinToolChains.cpp, should they just be static functions instead, perhaps with an extra argument for the `Darwin` instance? (This is genuinely a question; I'm not sure if that would be better.)

Comment (jrose-apple) [reply lib/Driver/ToolChains.h]: C++ makes it tricky to do that if the helper needs to access any private members, but if it doesn't then I agree.

File: lib/Driver/DarwinToolChains.cpp (hunk 2 above)
Comment (LucianoPAlmeida) [reply lib/Driver/DarwinToolChains.cpp]: Sure, just reverted :)

File: lib/Driver/ToolChains.h (hunk 3 above)
Comment (jrose-apple) [reply lib/Driver/ToolChains.h]: Heh. Brent meant the other kind of `static`, i.e. a local function in the .cpp file.

Comment (LucianoPAlmeida) [reply lib/Driver/ToolChains.h]: Ohh I see, not so hard to change.

**Comment by LucianoPAlmeida** [comment]
Still not sure if we could move validateSearchPaths. The other ones remaining seems general.

**Comment by jrose-apple** [comment]
Yeah, I think even though framework search paths aren't really used on other platforms, we still don't want to limit those to Darwin. Maybe the supported SDK is the only interesting thing.

(Also, watch those 4-space indents!)

**Comment by LucianoPAlmeida** [comment]
Sure,

> Yeah, I think even though framework search paths aren't really used on other platforms, we still don't want to limit those to Darwin. Maybe the supported SDK is the only interesting thing.
> 

Sure, just reverted that then :))

> (Also, watch those 4-space indents!)
Sorry, it is set as Xcode default :)

**Review by jrose-apple — APPROVED** [review]

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

File: lib/Driver/DarwinToolChains.cpp [hunk 4]
```diff
@@ -585,9 +585,45 @@ bool toolchains::Darwin::shouldStoreInvocationInDebugInfo() const {
   return false;
 }
 
+static void validateDeploymentTarget(const toolchains::Darwin& TC,
```
Comment (brentdax) [inline lib/Driver/DarwinToolChains.cpp]: Nit: We prefer the space before the `&`, not after.

**Review by beccadax — COMMENTED** [review]
One itty-bitty, teensy-weensy issue…

File: lib/Driver/DarwinToolChains.cpp (hunk 4 above)
Comment (LucianoPAlmeida) [reply lib/Driver/DarwinToolChains.cpp]: Sure, fixed :))

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
Thanks once more!

**Pull request merged**